### PR TITLE
[cli] misc enhancements

### PR DESCRIPTION
**Pull request by abtink**
This commit contains a group of smaller enhancements in `Cli` modules:

- Add and use `OutputLine()` helper method which outputs a formatted 
  string  line (appends newline `\r\n`).
- Use binary search when finding the command in `sCommands` array (also
  adds`static_assert` check to ensure `sCommands` array is sorted).
- Define `Command` struct as a `private` nested type.
- Use `enum` constants (replacing `#define`).
- Rename method to `OutputResult()` (from `AppendReult`) to harmonize
  with rest of a `Output{Something}()` methods.
- Remove comments at `#endif` when within 20 lines of `#if`

--------

I kept the commits separate for easier review but would suggest to squash them all before merge.

### CODE REVIEW

**Review by simonlingoogle — APPROVED** [review]
LGTM :+1: 
Happy to see code size reduced.

**Review by bukepo — APPROVED** [review]
LGTM:+1:

File: src/cli/cli.cpp [hunk 1]
```diff
@@ -4416,58 +4256,79 @@ void Interpreter::ProcessDiag(uint8_t aArgsLength, char *aArgs[])
...
-                mUserCommands[i].mCommand(aArgsLength - 1, &aArgs[1]);
-                break;
-            }
-        }
-
-        if (i == mUserCommandsLength)
-        {
-            AppendResult(OT_ERROR_INVALID_COMMAND);
+            mUserCommands[i].mCommand(aArgsLength - 1, &aArgs[1]);
+            ExitNow();
```
Comment (Irving-cl) [inline src/cli/cli.cpp:4326]: Change `FindCommand` so that it could also be applied to `UserCommands`?

**Review by Irving-cl — APPROVED** [review]
LGTM👍

File: src/cli/cli.cpp (hunk 1 above)
Comment (abtink) [reply src/cli/cli.cpp:4326]: We cannot easily extend `FindCommand()` to use for both  `sCommand` and `mUserCommands` arrays:
- Unlike the `sCommands` we cannot know if the `mUserCommands` list is sorted or not at compile time (it's given to us at run-time), so cannot always use binary search for it.
- Another point is that their underlying types are different (one uses `Command` struct and the other one uses `otCliCommand`), most notably their handler type `mCommand` are different:
   - `Command` uses an `Interpreter` method pointer 
   ```void (Interpreter::*mCommand)(uint8_t aArgsLength, char *aArgs[])``` 
   -  but `otCliCommand` uses a normal function pointer `void (*mCommand)(uint8_t aArgsLength, char *aArgs[])`

Comment (Irving-cl) [reply src/cli/cli.cpp:4326]: Got it. Thanks.

File: src/cli/cli.cpp [hunk 2]
```diff
@@ -485,17 +332,19 @@ void Interpreter::ProcessHelp(uint8_t aArgsLength, char *aArgs[])
     OT_UNUSED_VARIABLE(aArgsLength);
     OT_UNUSED_VARIABLE(aArgs);
 
+    static_assert(IsArraySorted(sCommands, OT_ARRAY_LENGTH(sCommands)), "Command list is not sorted");
```
Comment (jwhui) [inline src/cli/cli.cpp:335]: Can/should we apply these optimizations to sub-commands (e.g. `dataset`)?

Comment (abtink) [reply src/cli/cli.cpp:335]: Good idea. I think we should. Will add this later.

I am also considering (for future) to add a bunch of helper methods in `CmdLineParser` (arg lookup based on a table, string match, parse as number, etc). This way they can be shared/used by CLI and factory diags.

**Review by jwhui — APPROVED** [review]

**Pull request merged**